src/hotspot/share/memory/metaspace.hpp

```diff
@@ -113,8 +117,30 @@ class Metaspace : public AllStatic {
   static MetaWord* allocate(ClassLoaderData* loader_data, size_t word_size,
                             MetaspaceObj::Type type);
 
-  static bool contains(const void* ptr);
-  static bool contains_non_shared(const void* ptr);
+  static bool contains(const void* ptr) {
+    return is_in_shared_metaspace(ptr) || // in cds
+           is_in_class_space(ptr) ||      // in class space
+           is_in_nonclass_metaspace(ptr); // in one of the non-class regions?
```

Comment:
So the shared_metaspace region, which is the memory for CDS is *not* in class_space?  I suppose for deallocate this doesn't matter since this memory is never deallocated.

Reply:
By terminology, it is not in class space. 

Class space I (we? don't we all?) call the section that is governed by the metaspace allocator. The CDS region containing Klass is not in class space. It is, however, in the *narrow Klass decoding range*:

- Klass decoding range: [CompressedKlassPointers::base() ... CompressedKlassPointers::base() + CompressedKlassPointers::range() )
- that range contains the CDS klass range, followed (after a possible gap) by the traditional class space. If CDS is off, its just the class space.

Reply:
There had been a nice ASCII art somewhere, but I cannot find it anymore; maybe someone removed it.

Reply:
This comment should then be "or the CDS archive", since the function is contains() and includes the non-class metaspace.  That is, is_in_shared_metaspace() includes all of the CDS archive.  I don't think the CompressedKlassPointers:::range limit is in the middle of the CDS archive memory area, unless CDS was changed without me noticing to separate class and non-class metadata (I hope not).